Feature/add blank nodes and literals

UlrichMD opened the pull request:
For the moment the number of Blank nodes is set to the fixed value 4. @MichaelRoeder what is your thought on this? Should it be configurable or should we go with a fixed amount?

Commit:
added addBlankNode method to GrphBasedGraph

Commit:
Handle Blank Nodes in TripleCreator

Commit:
add test for blankNodes

MichaelRoeder commented on ldcbench.api/src/main/java/org/dice_research/ldcbench/graph/GrphBasedGraph.java:
```diff
@@ -253,4 +263,16 @@ public void setGraphIdOfNode(int nodeId, int graphId, int externalId) {
+        addNodes(nodeCount);
+        for (int i = blankNodesIndex; i < this.getNumberOfNodes(); i++) {
+            //Get a random Source Node
+            int sourceNode = (new Random()).nextInt(blankNodesIndex);
```
Randomly choosing a source node is fine. However, the way you implemented it has two drawbacks.
1. It is not possible to easily implement another way to choose a node.
2. It is impossible to repeat a run. Each component of ORCA receives a seed value. It is able to derive other seed values from this initial value. These other seed values can be used to initialize processes that are based on random numbers like this one.

MichaelRoeder commented on ldcbench.api/src/main/java/org/dice_research/ldcbench/rdf/TripleCreator.java:
```diff
@@ -12,4 +12,6 @@
 
     public Triple createTriple(int sourceId, int propertyId, int targetId, int targetExtId, int targetExtGraphId);
+
+    public Triple createTripleWithBlankNode(int sourceId, int propertyId, int targetId, int targetExtId, int targetExtGraphId);
```
Why this additional method? In the other creator classes, you simply added a boolean flag to the create method. Here, you propose a completely new method although it mainly does what the `createTriple` method already does :thinking:

UlrichMD replied:
the new method is there to differentiate the two cases and know when to set the blankNode param to true when calling createNode function. Another way of doing this could be adding an extra boolean param to the createTriple Method. But it would also change the way this method is used in the rest of the code.

MichaelRoeder replied:
Please apply the same idea as in this comment: https://github.com/dice-group/orca/pull/82#discussion_r744787842

MichaelRoeder commented on ldcbench.ckan-node/src/main/java/org/dice_research/ldcbench/nodes/ckan/simple/SimpleCkanComponent.java:
```diff
@@ -154,7 +154,7 @@ private void addCloudNode(int node) throws Exception {
         while (!success) {
             try {
-                addDataSource(tripleCreator.createNode(0, -1, -2, false).toString());
+                addDataSource(tripleCreator.createNode(0, -1, -2, false, false).toString());
```
You could have overloaded the method, i.e., one version is your newly implemented version of `createNode`. The other could have the "old" set of parameters and could simply assume that the blank node flag is false. This second method would then simply call your new method. This could have saved you the effort to change the calls everywhere. :thinking:

MichaelRoeder commented on ldcbench.data-generator/src/main/java/org/dice_research/ldcbench/benchmark/DataGenerator.java:
```diff
@@ -337,6 +337,12 @@ public void init() throws Exception {
             generator.generateGraph(avgDegree, numberOfEdges, seed, graph);
         }
 
+        if (type == Types.RDF_GRAPH_GENERATOR) {
+            //CHOOSE THE NUMBER OF BLANK NODE?? Currently using a fixed number of bNode
+            //DECIDE IF WE WANT BLANK NODES OR NOT ??
+            graph.addBlankNodes(3);
+        }
```
Please use a user-defined parameter as discussed. This would also have to be added to the benchmark.ttl file.

MichaelRoeder reviewed (changes requested):
Please add the parameter as discussed. Please have a look at the other comments as well.

Commit:
- Add blankNodeRatio parameter
- refactor createTriple Method
- update Readme
- use seed in addBlankNodes
- Add Test for createTripe with BlankNode

MichaelRoeder commented on ldcbench.api/src/main/java/org/dice_research/ldcbench/graph/GrphBasedGraph.java:
```diff
@@ -253,4 +263,17 @@ public void setGraphIdOfNode(int nodeId, int graphId, int externalId) {
         return graph.toString();
     }
+
+    @Override
+    public void addBlankNodes(int nodeCount, long seed) {
+        blankNodesIndex = this.getNumberOfNodes();
+        Random generator = new Random(seed);
+        addNodes(nodeCount);
+        for (int i = blankNodesIndex; i < this.getNumberOfNodes(); i++) {
+            //Get a random Source Node
+            int sourceNode = generator.nextInt(blankNodesIndex);
+            //Set the proper type?
+            addEdge(sourceNode, i, 0);
+        }
+    }
```
It is not the task of the graph to randomly add a blank node. This class is only responsible to store data about the graph and (as GraphBuilder instance) interact with the graph by adding and removing certain nodes or edges. 

However, the functionality you implemented above goes beyond that as you implement a certain algorithm to choose other nodes to which you can add the blank nodes. You also hard coded the property ID of the edge. All of this are things that are part of the overall generation algorithm implementation and shouldn't be in this Graph class.

Please note that the implementation seems to be fine. It is only in the wrong class ;)
Feel free to create a "BlankNodeAdder" with this piece of code. This class could then be used by all graph generator classes.

MichaelRoeder commented on ldcbench.api/src/main/java/org/dice_research/ldcbench/rdf/TripleCreator.java:
```diff
@@ -12,4 +12,9 @@
     public Triple createTriple(int sourceId, int propertyId, int targetId, int targetExtId, int targetExtGraphId);
+
+    /**
+     * Create a triple where the Object is a BlankNode
```
`... where the object might be a BlankNode`
would be more precise :wink:

MichaelRoeder commented on ldcbench.http-node/src/main/java/org/dice_research/ldcbench/nodes/utils/TripleIterator.java:
```diff
@@ -41,6 +41,10 @@ public Triple next() {
     }
 
     private Triple createTriple(int targetId, int propertyId) {
+        if (targetId >= graphs[datasetId].getBlankNodesIndex()) {
+            return tripleCreator.createTriple(nodeId, propertyId, targetId,
+                    graphs[datasetId].getExternalNodeId(targetId), graphs[datasetId].getGraphId(targetId), true);
+        }
         return tripleCreator.createTriple(nodeId, propertyId, targetId,
                 graphs[datasetId].getExternalNodeId(targetId), graphs[datasetId].getGraphId(targetId));
     }
```
That is fine, but why not simply
```
private Triple createTriple(int targetId, int propertyId) {
    return tripleCreator.createTriple(nodeId, propertyId, targetId,
                graphs[datasetId].getExternalNodeId(targetId), graphs[datasetId].getGraphId(targetId),
                targetId >= graphs[datasetId].getBlankNodesIndex());
```
Your `if` statement only decides whether the flag at the end is true or false. So you can directly use the condition of the if statement.

MichaelRoeder commented on ldcbench.api/src/main/java/org/dice_research/ldcbench/graph/Graph.java:
```diff
@@ -126,4 +126,11 @@ public default int incomingEdgeCount(int nodeId) {
      * @return the external ID of the node in the other graph.
      */
     public int getExternalNodeId(int nodeId);
+
+    /**
+     * Returns the ID of the first Blank Node of the graph
+     *
+     * @return the ID of the first Blank Node
+     */
+    public int getBlankNodesIndex();
```
Overall, the idea of using the ID to decide whether it is a blank node or not is good. However, maybe the interface of the Graph class would be much easier to use, if you add another method as follows:
```java
public default boolean isBlankNode(int nodeId) {
    return nodeId >= getBlankNodesIndex();
}
```
This would ease the usage of the Graph class because the user of the class does not have to know how the internal IDs work. They only want to know whether a node is a blank node or not :wink:

MichaelRoeder reviewed (commented):
A good PR. There are only some small enhancements left.

Commit:
fix review remarks

MichaelRoeder reviewed (approved):
LGTM

Commit:
remove unused imports
